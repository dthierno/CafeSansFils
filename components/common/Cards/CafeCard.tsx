import { View, Text, StyleSheet, Image, Pressable } from "react-native";
import { Circle } from "lucide-react-native";
import { router } from "expo-router";


import TYPOGRAPHY from "@/constants/Typography";
import COLORS from "@/constants/Colors";
import SPACING from "@/constants/Spacing";

type CafeCardProps = {
  status: "open" | "closing soon" | "closed";

  /** The name of the cafe */
  name: string;

  /** The location of the cafe */
  location: string;

  /** The price range of the cafe */
  priceRange: "$" | "$$" | "$$$";

  /** The rating of the cafe */
  rating: number;

  /** The image of the cafe */
  image?: string;

  /** The size of the card */
  size?: "medium" | "large";

  /** The slug of the cafe */
  slug?: string;
};

let cardDimensions = {
  medium: {
    width: 318,
    height: 180,
    image: require("@/assets/images/placeholder/imagemd.png"),
  },
  large: {
    width: 361,
    height: 210,
    image: require("@/assets/images/placeholder/imagexl.png"),
  },
};

/**
 * ## CafeCard Component
 *
 * A reusable card component that displays information about a cafe, including its name, location,
 * price range, rating, and status. The card also includes an image and supports navigation to the
 * cafe's details page when pressed.
 *
 * ### Example Usage
 *
 * ```tsx
 * <CafeCard
 *   status="open"
 *   name="Cozy Coffee"
 *   location="123 Coffee Lane"
 *   priceRange="$$"
 *   rating={4.5}
 *   image="https://example.com/image.jpg"
 *   size="large"
 *   slug="cozy-coffee"
 * />
 * ```
 *
 * @param {CafeCardProps} props - The props for the CafeCard component.
 */
export default function CafeCard({
  status,
  name,
  location,
  priceRange,
  rating,
  image,
  size = "medium",
  slug = "INVALID_SLUG",
}: CafeCardProps) {
  return (
    <Pressable
      onPress={() => router.push(`/cafe/${slug}`)}
      style={{ width: cardDimensions[size].width }}
      testID="button"
    >
      <View>
        <Image
          source={image ? { uri: image } : cardDimensions[size].image}
          width={cardDimensions[size].width}
          height={cardDimensions[size].height}
          style={{ borderRadius: SPACING["sm"] }}
          testID="image"
        />
        <Text
          style={[
            TYPOGRAPHY.body.small.bold,
            styles.priceRangeIcon,
            styles.rating,
          ]}
          testID="icon-button"
        >
          {rating}
        </Text>
      </View>
      <View style={styles.caption}>
        <View style={styles.cafeInfo}>
          <View style={styles.cafeInfoHeader}>
            <Text style={[TYPOGRAPHY.body.large.semiBold]}>{name}</Text>
            <Circle
              width={12}
              height={12}
              strokeWidth={1}
              color={
                status === "open"
                  ? COLORS.status.green
                  : status === "closing soon"
                  ? COLORS.status.orange
                  : COLORS.status.red
              }
              fill={
                status === "open"
                  ? COLORS.status.green
                  : status === "closing soon"
                  ? COLORS.status.orange
                  : COLORS.status.red
              }
              testID="tooltip-icon"
            />
          </View>
          <Text
            style={[TYPOGRAPHY.body.normal.semiBold, styles.cafeInfoLocation]}
          >
            {location}
          </Text>
        </View>
        <Text
          style={[TYPOGRAPHY.body.normal.base, styles.priceRangeIcon]}
          testID="icon-button"
        >
          {priceRange}
        </Text>
      </View>
    </Pressable>
  );
}

const styles = StyleSheet.create({
  caption: {
    justifyContent: "space-between",
    marginTop: SPACING["lg"],
    flexDirection: "row",
    alignItems: "center",
  },
  cafeInfo: {
    gap: SPACING["xxs"],
  },
  cafeInfoHeader: {
    flexDirection: "row",
    alignItems: "center",
    gap: SPACING["xs"],
  },
  cafeInfoLocation: {
    color: COLORS.black45,
  },
  priceRangeIcon: {
    backgroundColor: COLORS.lightGray,
    paddingHorizontal: SPACING.xs,
    paddingVertical: SPACING.xs,
    justifyContent: "center",
    borderRadius: 500,
  },
  rating: {
    backgroundColor: COLORS.white, // FIXME: Remove this line later
    position: "absolute",
    right: SPACING.sm,
    top: SPACING.sm,
  },
});
