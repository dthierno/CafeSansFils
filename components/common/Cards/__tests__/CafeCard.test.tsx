import React from "react";
import { render, screen, fireEvent } from "@testing-library/react-native";
import CafeCard from "@/components/common/Cards/CafeCard";
import { router } from "expo-router";
import COLORS from "@/constants/Colors";

jest.mock("expo-router", () => ({
  router: {
    push: jest.fn(),
  },
}));

describe("CafeCard Component", () => {
  const defaultProps = {
    status: "open" as "open" | "closing soon" | "closed",
    name: "Cozy Coffee",
    location: "123 Coffee Lane",
    priceRange: "$$" as "$" | "$$" | "$$$",
    rating: 4.5,
    slug: "cozy-coffee",
  };

  it("renders correctly with default props", () => {
    render(<CafeCard {...defaultProps} />);
    
    // Check if name is rendered
    expect(screen.getByText("Cozy Coffee")).toBeTruthy();

    // Check if location is rendered
    expect(screen.getByText("123 Coffee Lane")).toBeTruthy();

    // Check if price range is rendered
    expect(screen.getByText("$$")).toBeTruthy();

    // Check if rating is rendered
    expect(screen.getByText("4.5")).toBeTruthy();
  });

  it("renders the correct status icon color based on status", () => {
    const { rerender } = render(<CafeCard {...defaultProps} status="open" />);
    const greenIcon = screen.getAllByTestId("tooltip-icon");
    expect(greenIcon[0].props.fill).toBe(COLORS.status.green); // Replace with COLORS.status.green

    rerender(<CafeCard {...defaultProps} status="closing soon" />);
    const orangeIcon = screen.getAllByTestId("tooltip-icon");
    expect(orangeIcon[0].props.fill).toBe(COLORS.status.orange); // Replace with COLORS.status.orange

    rerender(<CafeCard {...defaultProps} status="closed" />);
    const redIcon = screen.getAllByTestId("tooltip-icon");
    expect(redIcon[0].props.fill).toBe(COLORS.status.red); // Replace with COLORS.status.red
  });

  it("navigates to the correct slug when pressed", () => {
    render(<CafeCard {...defaultProps} />);
    const pressable = screen.getByTestId("button");
    
    fireEvent.press(pressable);

    expect(router.push).toHaveBeenCalledWith("/cafe/cozy-coffee");
  });

  it("renders the fallback image when no image is provided", () => {
    render(<CafeCard {...defaultProps} image={undefined} size="medium" />);
    
    const image = screen.getByTestId("image");
    expect(image.props.source).toEqual(
      require("@/assets/images/placeholder/imagemd.png")
    );
  });

  it("renders the custom image when provided", () => {
    const customImage = "https://fakeimg.pl/600x400";
    render(<CafeCard {...defaultProps} image={customImage} />);
    
    const image = screen.getByTestId("image");
    expect(image.props.source.uri).toBe(customImage);
  });

  it("applies the correct card dimensions based on size", () => {
    const { rerender } = render(<CafeCard {...defaultProps} size="medium" />);
    expect(screen.getByTestId("button").props.style).toMatchObject({ width: 318 });

    rerender(<CafeCard {...defaultProps} size="large" />);
    expect(screen.getByTestId("button").props.style).toMatchObject({ width: 361 });
  });

  it("matches the snapshot", () => {
    const tree = render(<CafeCard {...defaultProps} />);
    expect(tree).toMatchSnapshot();
  });
});
