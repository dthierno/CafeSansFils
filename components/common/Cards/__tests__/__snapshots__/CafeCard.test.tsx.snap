// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CafeCard Component matches the snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "width": 318,
    }
  }
  testID="button"
>
  <View>
    <Image
      height={180}
      source={
        {
          "testUri": "../../../assets/images/placeholder/imagemd.png",
        }
      }
      style={
        {
          "borderRadius": 12,
        }
      }
      testID="image"
      width={318}
    />
    <Text
      style={
        [
          {
            "fontFamily": "Inter-Bold",
            "fontSize": 10,
          },
          {
            "backgroundColor": "#F2F2F7",
            "borderRadius": 500,
            "justifyContent": "center",
            "paddingHorizontal": 8,
            "paddingVertical": 8,
          },
          {
            "backgroundColor": "#FFFFFF",
            "position": "absolute",
            "right": 12,
            "top": 12,
          },
        ]
      }
      testID="icon-button"
    >
      4.5
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 20,
      }
    }
  >
    <View
      style={
        {
          "gap": 4,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 8,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "Inter-SemiBold",
                "fontSize": 14,
              },
            ]
          }
        >
          Cozy Coffee
        </Text>
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight={12}
          bbWidth={12}
          fill="#3E9C66"
          focusable={false}
          height={12}
          meetOrSlice={0}
          minX={0}
          minY={0}
          stroke="#3E9C66"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={1}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 12,
                "width": 12,
              },
            ]
          }
          testID="tooltip-icon"
          vbHeight={24}
          vbWidth={24}
          width={12}
          xmlns="http://www.w3.org/2000/svg"
        >
          <RNSVGGroup
            fill={
              {
                "payload": 4282293350,
                "type": 0,
              }
            }
            propList={
              [
                "fill",
                "stroke",
                "strokeWidth",
                "strokeLinecap",
                "strokeLinejoin",
              ]
            }
            stroke={
              {
                "payload": 4282293350,
                "type": 0,
              }
            }
            strokeLinecap={1}
            strokeLinejoin={1}
            strokeWidth={1}
          >
            <RNSVGCircle
              cx="12"
              cy="12"
              fill={
                {
                  "payload": 4282293350,
                  "type": 0,
                }
              }
              propList={
                [
                  "fill",
                  "stroke",
                  "strokeWidth",
                  "strokeLinecap",
                  "strokeLinejoin",
                ]
              }
              r="10"
              stroke={
                {
                  "payload": 4282293350,
                  "type": 0,
                }
              }
              strokeLinecap={1}
              strokeLinejoin={1}
              strokeWidth={1}
              testID="tooltip-icon"
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
      <Text
        style={
          [
            {
              "fontFamily": "Inter-SemiBold",
              "fontSize": 12,
            },
            {
              "color": "rgba(0, 0, 0, 0.45)",
            },
          ]
        }
      >
        123 Coffee Lane
      </Text>
    </View>
    <Text
      style={
        [
          {
            "fontFamily": "Inter-Regular",
            "fontSize": 12,
          },
          {
            "backgroundColor": "#F2F2F7",
            "borderRadius": 500,
            "justifyContent": "center",
            "paddingHorizontal": 8,
            "paddingVertical": 8,
          },
        ]
      }
      testID="icon-button"
    >
      $$
    </Text>
  </View>
</View>
`;
